### PR TITLE
[BIOIN-1629] Helper script for de novo refinement

### PR DESCRIPTION
**Pull request by ilyasoifer**

### CODE REVIEW

File: ugvc/pipelines/denovo_recalibrated_qualities.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import sys
+
+from ugvc.joint import denovo_refinement
+
+if __file__ == "__main__":
+    denovo_refinement.main(sys.argv[1:])
+
+
+def run(argv: list):
+    "Add recalibrated through somatic calling qualities to the annotated de novo VCF"
```
Comment (doron-st) [inline ugvc/pipelines/denovo_recalibrated_qualities.py]: The grammer doesn't sound correct to me - can we clarify this sentence?

File: ugvc/joint/denovo_refinement.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    paternal_vcfs : dict
+        dictionary with keys: names of the children and values: paths to the paternal VCFs
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe corresponding to the combined parental VCFs.
+        The columns are of the form {child id}-mother, {child id}-father
+    """
+    maternal_dfs = [(x + "-mother", vcftools.get_vcf_df(maternal_vcfs[x])) for x in tqdm.tqdm(maternal_vcfs)]
```
Comment (doron-st) [inline ugvc/joint/denovo_refinement.py:31]: do we add a dataframe and a string here, and it works?

File: ugvc/joint/denovo_refinement.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    maternal_df = pd.concat(dict(maternal_dfs), axis=1)
+    paternal_dfs = [(x + "-father", vcftools.get_vcf_df(paternal_vcfs[x])) for x in tqdm.tqdm(paternal_vcfs)]
+    paternal_df = pd.concat(dict(paternal_dfs), axis=1)
+    parent_df = pd.concat([maternal_df, paternal_df], axis=1)
+    return parent_df
+
+
+def add_paternal_qualities_to_denovo_vcf(denovo_vcf: str, parental_vcf_df: pd.DataFrame) -> pd.DataFrame:
```
Comment (doron-st) [inline ugvc/joint/denovo_refinement.py]: typo - parental

File: ugvc/joint/denovo_refinement.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        Dataframe with recalibrated qual (pair_qual) added to each denovo call in the denovo_vcf
+        The pair qual is the minimal qual of the variant in the two child/parent vcfs
+    """
+    denovo_df = vcftools.get_vcf_df(denovo_vcf)
+    denovo_df = pd.concat([denovo_df, parental_vcf_df.xs("qual", axis=1, level=1)], axis=1)
+    df_denovo_exp = denovo_df.explode("hiconfdenovo").explode("loconfdenovo")
+    df_denovo_exp["denovosample"] = np.where(
+        pd.isnull(df_denovo_exp["hiconfdenovo"]), df_denovo_exp["loconfdenovo"], df_denovo_exp["hiconfdenovo"]
+    )
+    called_samples = set(x[:-7] for x in df_denovo_exp.columns if x.endswith("mother"))
```
Comment (doron-st) [inline ugvc/joint/denovo_refinement.py]: I don't understand this code. but isn't the assymetry problematic? we look if it's called in the mother specifically?

**Review by doron-st — APPROVED** [review]
minor comments

File: ugvc/joint/denovo_refinement.py (hunk 2 above)
Comment (ilyasoifer) [reply ugvc/joint/denovo_refinement.py:31]: maternal_vcfs is a list of strings

File: ugvc/joint/denovo_refinement.py (hunk 4 above)
Comment (ilyasoifer) [reply ugvc/joint/denovo_refinement.py]: Rewrote for clarity, thanks!

File: ugvc/pipelines/denovo_recalibrated_qualities.py (hunk 1 above)
Comment (ilyasoifer) [reply ugvc/pipelines/denovo_recalibrated_qualities.py]: done

**Review by gilhornung — APPROVED** [review]
lgtm

**Pull request merged**